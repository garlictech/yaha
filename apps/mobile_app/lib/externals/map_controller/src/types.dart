/// The different map tile layers available
enum TileLayerType {
  /// Normal layer
  normal,

  /// Topographic layer
  topography,

  /// Monochrome layer
  monochrome,

  /// Hike layer
  hike,

  /// Custom tile layer
  custom
}
