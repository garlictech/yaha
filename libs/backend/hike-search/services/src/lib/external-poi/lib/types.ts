import { YahaApi } from '@yaha/gql-api';

export interface ExternalPoi {
  externalId: string;
  location: {
    lat: number;
    lon: number;
  };
  description: YahaApi.Description;
  type?: string;
  infoUrl?: string;
  address?: string;
  openingHours?: string;
  phoneNumber?: string;
  elevation?: number;
}

export interface ExternalImage {
  externalId: string;
  location: {
    lat: number;
    lon: number;
  };
  original: string;
  card: string;
  thumbnail: string;
}

export enum OsmPoiTypes {
  natural = 'natural',
  amenity = 'amenity',
  publicTransport = 'public_transport',
  emergency = 'emergency',
  historic = 'historic',
  leisure = 'leisure',
  manMade = 'man_made',
  military = 'military',
  shop = 'shop',
  tourism = 'tourism',
}

export interface GoogleApiConfig {
  apiKey: string;
}
