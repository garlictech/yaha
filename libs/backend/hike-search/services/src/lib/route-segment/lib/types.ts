import { YahaApi } from '@yaha/gql-api';
import { Position } from '@turf/helpers';
import { PathType } from '../../geometry';

export enum EBuffer {
  SMALL = 'small',
  BIG = 'big',
}

export const emptyRouteTotals = {
  distance: 0,
  uphill: 0,
  downhill: 0,
  averageTime: 0,
  score: 0,
  currentTime: 0,
};

export type RouteTotals = typeof emptyRouteTotals;

export interface RouteSegment extends RouteTotals {
  coordinates: Position[];
  difficulty: number;
  endPoint: YahaApi.Point;
  startPoint: YahaApi.Point;
  geojsonFeature: PathType;
}
