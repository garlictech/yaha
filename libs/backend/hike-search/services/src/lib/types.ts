import { Observable } from 'rxjs';
import { Either } from 'fp-ts/lib/Either';

export interface QueryData {
  action: string;
  stringifiedData: string;
}

export interface GraphqlCRUDApi<
  CREATE_TYPE,
  ENTITY_TYPE extends { id: string }
> {
  add(entity: CREATE_TYPE): Observable<Either<any, ENTITY_TYPE>>;
  delete(id: string): Observable<Either<any, boolean>>;
  getById(id: string): Observable<Either<any, ENTITY_TYPE>>;
  getWithQuery(query: QueryData): Observable<Either<any, ENTITY_TYPE[]>>;
  update(
    id: string,
    update: Partial<ENTITY_TYPE>
  ): Observable<Either<any, ENTITY_TYPE>>;
  createMultipleItems(
    input: CREATE_TYPE[]
  ): Observable<Either<any, ENTITY_TYPE[]>>;
  updateMultipleItems(
    input: Partial<ENTITY_TYPE>[]
  ): Observable<Either<any, ENTITY_TYPE[]>>;
  getMultipleItemsById(ids: string[]): Observable<Either<any, ENTITY_TYPE[]>>;
}

export type DataServiceActions =
  | 'resolveByMultipleKeys'
  | 'createMultipleItems';
