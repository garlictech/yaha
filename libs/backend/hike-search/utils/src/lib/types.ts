import {
  PoiSourceObject,
  TextualDescription,
  CreateImageInput,
} from '@bit/garlictech.universal.gtrack.graphql-api';

export type PoiSearchOutputType = (ExternalPoi | CreateImageInput)[];

interface ExternalPoiType {
  lat: number;
  lon: number;
  elevation: -1;
  id: '';
  types?: string[];
  sourceObject: PoiSourceObject[];
  description: TextualDescription[];
  address?: string;
  phoneNumber?: string;
  openingHours?: string;
}

export type ExternalPoi = Readonly<ExternalPoiType>;

export enum OsmPoiTypes {
  natural = 'natural',
  amenity = 'amenity',
  publicTransport = 'public_transport',
  emergency = 'emergency',
  historic = 'historic',
  leisure = 'leisure',
  manMade = 'man_made',
  military = 'military',
  shop = 'shop',
  tourism = 'tourism',
}

export interface GoogleApiConfig {
  apiKey: string;
}
