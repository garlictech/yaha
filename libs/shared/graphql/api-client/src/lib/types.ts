export interface CrudApiConfig extends Record<string, unknown> {
  aws_appsync_graphqlEndpoint: string;
  aws_appsync_region: string;
  aws_user_pools_id: string;
  aws_user_pools_web_client_id: string;
  aws_appsync_apiKey: string;
}

export interface AmplifyModel {
  createdAt?: number;
  updatedAt?: number;
}

export interface StorageResponse {
  key: string;
}
