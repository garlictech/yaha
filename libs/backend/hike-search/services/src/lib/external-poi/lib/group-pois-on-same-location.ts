import { YahaApi } from '@yaha/gql-api';
import { pipe } from 'fp-ts/lib/function';
import {
  cloneDeep,
  concat as _concat,
  filter,
  flatten,
  intersection,
  isEmpty,
  isEqual,
  map as _map,
  pull,
  pullAll,
  slice,
  uniq,
  uniqWith,
  without,
} from 'lodash';
import * as fp from 'lodash/fp';
import { approximateDistance } from '../../geometry';
import * as R from 'ramda';

function indexPoiPairsByDistance<
  POINT extends { location: { lat: number; lon: number } },
>(points: POINT[]): { distance: number; point1: POINT; point2: POINT }[] {
  const indexedPairsbyDistance = _map(points, (point1, index) => {
    const processedPoisRemoved = slice(points, index + 1);
    return _map(processedPoisRemoved, point2 => ({
      distance: approximateDistance(point1.location, point2.location),
      point1,
      point2,
    }));
  });

  return fp.sortBy(['distance'], fp.flatten(indexedPairsbyDistance));
}

/*const mergeSourceObjects = (
  pois: YahaApi.Poi[],
  targetPoi: YahaApi.Poi,
): YahaApi.PoiSourceObject[] =>
  fp.flow(
    fp.map('sourceObject'),
    fp.flatten,
    fp.concat(targetPoi.sourceObject),
    fp.reverse,
    fp.uniqWith(fp.isEqual),
  )(pois);
*/
export const mergePois = (
  pois: YahaApi.Poi[],
  targetPoi: YahaApi.Poi,
): YahaApi.Poi => ({
  ...targetPoi,
  sourceObject: targetPoi.sourceObject,
  //sourceObject: mergeSourceObjects(pois, targetPoi),
  description: uniqWith(
    uniqWith(
      flatten(_concat(targetPoi.description, fp.map('description')(pois))),
      isEqual,
    ),
    (desc1, desc2) => desc1.languageKey === desc2.languageKey,
  ),
  types: filter(
    uniq(flatten(_concat(targetPoi.types, fp.map('types')(pois)))),
    item => !!item,
  ),
});

export function groupPoisOnSameLocation(minimalDistanceInMeters: number) {
  return (pois: YahaApi.Poi[]): YahaApi.Poi[] => {
    let newPois: YahaApi.Poi[] = cloneDeep(pois);
    // create poi pairs indexed with their distance
    const indexedPairs = indexPoiPairsByDistance(newPois);
    let pairsCloseToEachOther = filter(
      indexedPairs,
      pair => pair.distance < minimalDistanceInMeters,
    );

    const findAndMergeClosePois = () => {
      const targetPoi = pairsCloseToEachOther[0].point1;
      let pickedPois = [targetPoi, pairsCloseToEachOther[0].point2];
      let pickedPairs = [pairsCloseToEachOther[0]];
      const wrapperFn = (points: YahaApi.Poi[]) =>
        indexPoiPairsByDistance(points);
      let pairsInRelationWithAPicked: ReturnType<typeof wrapperFn>;

      do {
        pairsInRelationWithAPicked = filter(
          pairsCloseToEachOther,
          pointToCheck =>
            !isEmpty(
              intersection(pickedPois, [
                pointToCheck.point1,
                pointToCheck.point2,
              ]),
            ),
        );

        pickedPairs = uniq(pickedPairs.concat(pairsInRelationWithAPicked));

        pickedPois = pipe(
          pairsInRelationWithAPicked.map(pair => [pair.point1, pair.point2]),
          fp.flatten,
          R.concat(pickedPois),
          fp.uniq,
        );

        pairsCloseToEachOther = pullAll(
          pairsCloseToEachOther,
          pairsInRelationWithAPicked,
        );
      } while (!isEmpty(pairsInRelationWithAPicked));

      const poisToMerge = without(pickedPois, targetPoi);

      if (!isEmpty(poisToMerge)) {
        const mergedPoi = mergePois(poisToMerge, targetPoi);
        newPois.push(mergedPoi);
        newPois = pull(newPois, targetPoi);
      }

      newPois = pullAll(newPois, poisToMerge);
    };

    while (!isEmpty(pairsCloseToEachOther)) {
      findAndMergeClosePois();
    }

    return newPois;
  };
}
