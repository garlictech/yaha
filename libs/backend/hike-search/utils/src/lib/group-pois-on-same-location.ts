import { approximateDistance } from '@bit/garlictech.universal.gtrack.geometry';
import {
  Poi,
  PoiSourceObject,
} from '@bit/garlictech.universal.gtrack.graphql-api';
import {
  cloneDeep,
  concat as _concat,
  filter,
  flatten,
  intersection,
  isEmpty,
  isEqual,
  map as _map,
  omit,
  pull,
  pullAll,
  slice,
  uniq,
  uniqWith,
  without,
} from 'lodash';
import * as fp from 'lodash/fp';
import { ExternalPoi } from './types';

function indexPoiPairsByDistance<POINT extends { lat: number; lon: number }>(
  points: POINT[],
): { distance: number; point1: POINT; point2: POINT }[] {
  const indexedPairsbyDistance = _map(points, (point1, index) => {
    const processedPoisRemoved = slice(points, index + 1);
    return _map(processedPoisRemoved, point2 => ({
      distance: approximateDistance(point1, point2),
      point1,
      point2,
    }));
  });

  return fp.sortBy(['distance'], fp.flatten(indexedPairsbyDistance));
}

type PoiGroupingInputType = Readonly<ExternalPoi> | Poi;
type PoiGroupingOutputType = ExternalPoi | Poi;

const removeUnusedGraphqlFields = fp.map(
  fp.flow(fp.omit(['__typename']), fp.omitBy(fp.isNull)),
);

const mergeSourceObjects = (
  pois: PoiGroupingInputType[],
  targetPoi: PoiGroupingOutputType,
): PoiSourceObject[] =>
  fp.flow(
    fp.map('sourceObject'),
    fp.flatten,
    fp.concat(targetPoi.sourceObject),
    fp.reverse,
    removeUnusedGraphqlFields,
    fp.uniqWith(fp.isEqual),
  )(pois);

export const mergePois = (
  pois: PoiGroupingInputType[],
  targetPoi: PoiGroupingOutputType,
): PoiGroupingOutputType => ({
  ...omit(targetPoi, ['sourceObject', 'description', 'types']),
  address: fp.isNull(targetPoi.address) ? undefined : targetPoi.address,
  sourceObject: mergeSourceObjects(pois, targetPoi),
  description: uniqWith(
    uniqWith(
      flatten(_concat(targetPoi.description, fp.map('description')(pois))),
      isEqual,
    ),
    (desc1, desc2) => desc1.languageKey === desc2.languageKey,
  ),
  types: filter(
    uniq(flatten(_concat(targetPoi.types, fp.map('types')(pois)))),
    item => !!item,
  ),
});

export function groupPoisOnSameLocation<POI extends PoiGroupingInputType>(
  minimalDistanceInMeters: number,
) {
  return (pois: POI[]): PoiGroupingOutputType[] => {
    let newPois: PoiGroupingOutputType[] = cloneDeep(pois);
    // create poi pairs indexed with their distance
    const indexedPairs = indexPoiPairsByDistance(newPois);
    let pairsCloseToEachOther = filter(
      indexedPairs,
      pair => pair.distance < minimalDistanceInMeters,
    );

    const findAndMergeClosePois = () => {
      const targetPoi = pairsCloseToEachOther[0].point1;
      let pickedPois = [targetPoi, pairsCloseToEachOther[0].point2];
      let pickedPairs = [pairsCloseToEachOther[0]];
      let pairsInRelationWithAPicked;

      do {
        pairsInRelationWithAPicked = filter(
          pairsCloseToEachOther,
          pointToCheck =>
            !isEmpty(
              intersection(pickedPois, [
                pointToCheck.point1,
                pointToCheck.point2,
              ]),
            ),
        );

        pickedPairs = uniq(pickedPairs.concat(pairsInRelationWithAPicked));
        pickedPois = uniq(
          flatten(
            pickedPois.concat(
              pairsInRelationWithAPicked.map(pair => [
                pair.point1,
                pair.point2,
              ]),
            ),
          ),
        );

        pairsCloseToEachOther = pullAll(
          pairsCloseToEachOther,
          pairsInRelationWithAPicked,
        );
      } while (!isEmpty(pairsInRelationWithAPicked));

      const poisToMerge = without(pickedPois, targetPoi);

      if (!isEmpty(poisToMerge)) {
        const mergedPoi = mergePois(poisToMerge, targetPoi);
        newPois.push(mergedPoi);
        newPois = pull(newPois, targetPoi);
      }

      newPois = pullAll(newPois, poisToMerge);
    };

    while (!isEmpty(pairsCloseToEachOther)) {
      findAndMergeClosePois();
    }

    return newPois;
  };
}
