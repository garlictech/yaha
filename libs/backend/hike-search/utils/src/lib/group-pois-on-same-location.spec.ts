import { groupPoisOnSameLocation } from './group-pois-on-same-location';
import { ExternalPoi } from './types';

test('Filter pois on same location', () => {
  const fixture = [
    {
      id: 1,
      lat: 1,
      lon: 1,
      sourceObject: [
        {
          objectId: 1,
        },
      ],
      types: ['t1'],
      description: {
        languageKey: 'lan1',
        summary: 'summary 1',
      },
    },
    {
      id: 2,
      lat: 1,
      lon: 1,
      types: ['t2.1', 't2.2'],
      sourceObject: [
        {
          objectId: 2,
        },
      ],
      description: {
        languageKey: 'lan1',
        summary: 'summary 2',
      },
    },
    {
      id: 3,
      lat: 2,
      lon: 2,
      sourceObject: [
        {
          objectId: 3,
        },
      ],
      description: {
        languageKey: 'lan3',
        summary: 'summary 3',
      },
    },
    {
      id: 4,
      lat: 1,
      lon: 1,
      sourceObject: [
        {
          objectId: 4,
        },
      ],
      description: {
        languageKey: 'lan4',
        summary: 'summary 4',
      },
    },
  ] as unknown as ExternalPoi[];

  expect(groupPoisOnSameLocation(50)(fixture)).toMatchSnapshot();
});
