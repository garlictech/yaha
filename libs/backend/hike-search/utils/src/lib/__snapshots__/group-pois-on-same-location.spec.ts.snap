// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter pois on same location 1`] = `
Array [
  Object {
    "description": Object {
      "languageKey": "lan3",
      "summary": "summary 3",
    },
    "id": 3,
    "lat": 2,
    "lon": 2,
    "sourceObject": Array [
      Object {
        "objectId": 3,
      },
    ],
  },
  Object {
    "address": undefined,
    "description": Array [
      Object {
        "languageKey": "lan1",
        "summary": "summary 1",
      },
      Object {
        "languageKey": "lan4",
        "summary": "summary 4",
      },
    ],
    "id": 1,
    "lat": 1,
    "lon": 1,
    "sourceObject": Array [
      Object {
        "objectId": 4,
      },
      Object {
        "objectId": 2,
      },
      Object {
        "objectId": 1,
      },
    ],
    "types": Array [
      "t1",
      "t2.1",
      "t2.2",
    ],
  },
]
`;
